epicsdbbuilder/const_array.py

```diff
@@ -0,0 +1,87 @@
+from decimal import Decimal
+from .recordbase import quote_string
+from .parameter import Parameter
+
+__all__ = ['ConstArray']
+
+
+class ConstArray:
+    """Constant Link Values. EPICS Base 3.16.1 and above.
+
+    Example: PY Source
+    ------------------
+    `r = records.lsi('r', INP=ConstArray(['Plain String not DBLINK']))`
+
+    Example: Generated DB
+    ---------------------
+    `field(INP, ["Plain String not DBLINK"])`
+    """
+
+    def __init__(self, value):
+        """Constructor.
+
+        Parameters
+        ----------
+        value : iterable
+            Iterable which can provide a homogeneous non-empty list of values.
+        """
+        self.__value = self._sanitize(value)
+
+    def _sanitize(self, raw_value):
+        # ConstArray allows iterable only.
+        value_list = list(raw_value)
+
+        # EPICS 7.0.3.1 does not consider "[]" as constant.
+        assert len(value_list) > 0, \
+            f'ConstArray: Empty iterable is not allowed.'
+
+        # EPICS does not allow mixing of strings and numbers.
+        numbers = False
+        strings = False
+        valid_types = (Parameter, str, int, float, bool, Decimal)
+        for index, value in enumerate(value_list):
+            assert isinstance(value, valid_types), \
+                f'ConstArray: expects a string or parameter as element' \
+                f' but an element at the index {index} is {type(value)}.'
+
+            if isinstance(value, (Parameter, str)):
```

Comment:
There's a small bug here!  A `Parameter` will be of unpredictable type, so you can't treat it as either a string _or_ a number, so I think you need a third test, perhaps along the lines of:
```python
if isinstance(value, Parameter):
    pass
elif isinstance(value, str):
    ... test for strings...
else:
    ... test for numbers...
```
Or else: rename `valid_types` to `number_types` and take out `Parameter`, `str` from the list, and test for that in the last clause:
```python
if isinstance(value, Parameter):
    pass  # Always accept Parameters
elif isinstance(value, str):
    ... string ...
elif isintance(value, number_types):
    ... number ...
else:
    assert False, ... expects a string number or parameter ...
```

Reply:
I am afraid I can treat it as unpredictable until certain point. And that point is formatting. Right now, the parameter will be treated as a string (e.g. `"$(X)"`) so that it can be used with string record types (e.g. lsi). If we take it out of the condition, we might end up with `[1, "$(X)", 2]` on the output. This will not work right in case of non-string types (see table below). A workaround could be to treat a parameter as a number and allow the user to supply it in a form of a string (e.g. `$(X)`) if it is meant to be treated as a string. Yet, I am not sure how we can tell a reference to a parameter X apart from a string `$(X)` without additional information. This assume we want to treat a string as a string without any ambiguity.

I think, there is a chance to archive reasonable behavior with some guesswork. Yet, this is case-by-case situation. In case of string record type (lsi/lso, stringin/stringout), a parameter is meant to be a string. Incase of arrays, another field might give a hint. While this could be created, it will not be that quick. Therefore, I focused on getting some functionality without any loose ends or gaps first. Therefore, I added more restrictive rule (i.e. parameters are strings). Definitely, we can remove the support for parameters if we want to. What behavior would you like to prefer?

Here are some results of experimenting with various record types on EPICS 7.0.3.1. Given the style EPICS treats input, it is difficult to tell the intention from an undefined or unintended behavior (a side-effect). The table below shows a record type with relevant fields, a content of the field INP, and result of caget:
| record type | INP | caget | loading DB |
| --- | --- | --- | --- |
| waveform (FTVL=DOUBLE) | `[1, 2, 3]` | `3 1 2 3` |  |
| waveform (FTVL=DOUBLE) | `["1", "2", "3"]` | `3 0 0 0` | warning `dbConvertJSON: String provided, numeric value(s) expected` |
| waveform (FTVL=DOUBLE) | `[1, "2", 3]` | `3 0 0 0` | warning `dbConvertJSON: String provided, numeric value(s) expected` |
| waveform (FTVL=STRING) | `["1", "2", "3"]` | `3 1 2 3` |  |
| waveform (FTVL=STRING) | `[1, "2", 3]` | `3 1 2 3` |  |
| lsi | `[123]` | `` (empty string) |  |
| lsi | `["$(X)"]` | `ValueOfX` |  |
| lsi | `[$(X)]` | `` (empty string) | syntax error while loading |

Reply:
What does the db loader do when you give it `[1, $(X), 3]` with `X=2` specified?  (I don't have an up to date version of EPICS to hand to test with).  I would have expected that to expand to `[1, 2, 3]`.

However, I see the problem when it comes to expanding the parameter itself: at the moment you treat it as a string and expand it to `"$(X)"`; without knowing the underlying record type (which we can't rely on knowing, particularly if it depends on the value of another field) it's impossible to reliably choose between `$(X)` and `"$(X)"`!

Hmph.  Guess we can't fix this easily, and I'm not sure that it would be a problem in most applications.

Reply:
In case of `X=2` and `[1, $(X), 2]`, you will get `[1, 2, 3]`.

Regarding `$(X)` and `"$(X)"`. I think the key would be to identify when the decision is not possible to be made. In such a scenario, the routine should fail and the user should be allowed to add a hint (e.g. `ConstArray([AsString(par)])`). As long as the logging is right, the user will not be frustrated: the log will help the user to move forward without googling or guessing. Yet, I see this as a topic for another pull request. Right now, we can either decide on how we want to treat the parameter (and add all the checks to make sure the we generate valid output) or deny use of parameters. I think I use a parameter with a record type lsi. Thus, I preferred treating the parameter as a string.

Reply:
Yes, I agree with all that, except I think I'd add the type hint as an optional extra parameter to `ConstArray`, eg `ConstArray([str_par], as_string=True)` and `ConstArray([num_par], as_string=False)` ... and then we can worry whether the parameter should be `as_string`, `as_num`, or `type`, and whether the keyword is optional.

As you say, let's do this another time.